### PR TITLE
[Android] Add W3C Presentation API impl.

### PR DESCRIPTION
**Pull request by yuankaining**
BUG=[XWALK-4646](https://crosswalk-project.org/jira/browse/XWALK-4646)

### Background
W3C Presentation API is defined to enable JavaScript developers to show a new web frame (with a specified URL) on an available 2nd display (e.g. a 2nd computer monitor, a Miracast featured TV).

### Demo
An online [demo](https://storage.googleapis.com/presentation-api/index.html) (Note: doesn't work on a normal browser)
Or, a simplified JS source demo

    var request = new PresentationRequest("http://example.com/page-to-show.html");
    request.getAvailability().then(function(avail) {
      if (avail.value) {
        request.start().then(...)
      }
    });

### This PR
There was a previous impl. for W3C Presentation API on Android, written by @jondong , which was using xwalk extensions mechanism to provide the API interface and functionality. So the calling graph would be like:

`[JavaScript call] o--|UA boundary|--> [bolt-on Java + JavaScript as an extension] o--|UA Boundary|--> [Android system]`

And this PR is using the API interface provided by Blink, so the new calling graph:

`[JavaScript call] o--|UA boundary|--> [Blink API Interface] o--> [Chromium internal service API ] o-->  [built-in Java & C++ impl] o--|UA Boundary|--> [Android system]`

The reason that we want to migrate to Blink API is obvious: we don't want to put much effort to keep up with every API changes of that changing spec (it did changed a lot in a few months). By doing this, we can leverage Blink authors' effort and just maintain the impl. of internal Chromium service interface, which is the easier part, e.g. for the W3C API changes happened in recent months, we didn't need to do anything besides renaming a few equivalent constants.

### Related PR
@pozdnyakov had done [a similar PR](https://github.com/crosswalk-project/crosswalk/pull/3400) for to impl. W3C Presentation API for Windows platform. The difference is only the platform.

### CODE REVIEW

**Comment by rakuco** [comment]
Please be more verbose in both your pull request message and commit message. We need to figure out from those what is being done, how it changes the existing implementation and, most importantly, _why_.

**Comment by yuankaining** [comment]
@jondong @pozdnyakov I've just created the PR for Android W3C Presentation API impl. Can you please review the change? (I'm supposing the review happens in a PR, isn't it?)

The current code share the similar design/code with Windows impl. , which are being unified soon... @pozdnyakov I'll start drafting the refactoring proposal, when you have time to do it? Thanks.

**Comment by yuankaining** [comment]
Hi @rakuco, roger that. Will do (please allow a few hours buffer, I'm leaving office)

**Comment by yuankaining** [comment]
Now github.com is back :)

Not sure why trybot is fail. looking into it right now.

**Comment by yuankaining** [comment]
The failure is due to coding style... is there a local tool that can help check that?

**Comment by halton** [comment]
@yuankaining please find the lint tool under `src/xwalk/tools/lint.py`

**Comment by yuankaining** [comment]
@halton Yep, thanks. Coding issues were fixed and now I'm running a build to test it before pushing my commit.

**Comment by halton** [comment]
@jondong @pozdnyakov could you prioritize your time to review this PR, it is a P1 blocker.

**Comment by jondong** [comment]
@yuankaining Thanks for the PR, would you please squash these 2 commits into one and force update your branch? We'd better make sure that every commit in trunk passes the trybot. Looking into your implementation...

**Comment by yuankaining** [comment]
Thanks to @jondong for the input, is there any other issues need to be taken care of? If not, I'll start to work on those ones :)

**Comment by yuankaining** [comment]
Thanks @jondong for the review, all issues you've found are fixed now.

**Comment by yuankaining** [comment]
@pozdnyakov Hi, Mikhail, what do you think of this change? Do you have time to review it? Thanks in advance :)

File: runtime/browser/xwalk_presentation_service_delegate_android.cc [hunk 1]
```diff
@@ -0,0 +1,742 @@
...
+    callback.Run(session, "");
+
+  } else {
+    LOG(ERROR) << "XWalkPresentationHost instance not found";
+  }
+}
+
+void PresentationSession::Close() {
+  if ( auto presentation_api = XWalkPresentationHost::GetInstance() ) {
+    presentation_api->closePresentation(this->render_process_id_,
```
Comment (pozdnyakov) [inline runtime/browser/xwalk_presentation_service_delegate_android.cc]: nit: this-> is not needed

File: runtime/browser/xwalk_presentation_service_delegate_android.cc [hunk 2]
```diff
@@ -0,0 +1,742 @@
...
+      this->render_frame_id_);
```
Comment (pozdnyakov) [inline runtime/browser/xwalk_presentation_service_delegate_android.cc]: ditto.

**Comment by pozdnyakov** [comment]
The code looks OK but I was hoping you would share the existing code of the Windows implementation rather than copy it.. 
if the issue is urgent we can fix it later, otherwise it should be re-written to a class hierarchy like (
generic  XWalkPresentationServiceDelegate and  inheriting                                        XWalkPresentationServiceDelegateAndroid  XWalkPresentationServiceDelegateWin)

**Comment by darktears** [comment]
+1, there is a lot that can be shared and that something I'm trying to do with new stuff I'm adding. Let's go that way please. This is valid for future contribution, and I'm sure today there is stuff that could be refactored in the Android port to be shared with desktop.

**Comment by yuankaining** [comment]
@pozdnyakov Yes, they should share the same copy of code, we've talked about this on email and I already have a plan almost done for your review. The reason I didn't do it now is that I wanted to do it in another PR so that I can close this P1 bug ASAP. This bug was not important at that time and I thought I had plenty of time before it was suddenly marked as P1 :)   I'll send out the plan to you by EOD today.

**Comment by yuankaining** [comment]
@pozdnyakov Hi, Mikhail, the refactoring proposal was just sent to you via email...

And from my understanding of your above comment, this PR is ready to merge, is this information correct? I'm still not familiar with the process yet: who will be the one to merge this PR? Are you the one? Thanks :)

**Comment by pozdnyakov** [comment]
> And from my understanding of your above comment, this PR is ready to merge, is this information correct? 
LGTM considering the promise to share the repeated code

**Comment by halton** [comment]
LGTM for the android part, plus @pozdnyakov 's LGTM for the runtime part, @darktears @kenchris could you help to merge this PR? 

After it landed, @yuankaining will submit PR to crosswalk-18 branch to fix the P1 blocker .

**Comment by baleboy** [comment]
Please make sure there is a jira issue to track the removal of duplicated code, assigned and targeting Crosswalk 19.

@halton we haven't yet branched 18, so if this lands today or tomorrow it will be in 18 automatically.

**Comment by rakuco** [comment]
@yuankaining: thanks for making the PR message much more detailed than the original version, it's appreciated. Next time, please remember to update your commit message as well (`git commit --amend`); the commit that was merged still only says "BUG=XWALK-4646" without your great (in all senses :-) explanation message in the pull request.

**Comment by yuankaining** [comment]
@rakuco the commit message is actually this:

    commit 915b603793604d6fe77758368f5763f936579ecf
    Author: Kenny Yuan <kaining.yuan@intel.com>
    Date:   Fri Jan 29 19:42:50 2016 +0800

        [Android] Add W3C Presentation API impl.
    
        BUG=XWALK-4646

I'm following other colleagues's commit message style -- one line of brief message, one line of JIRA bug ID. Is this style welcomed? Let me know if there is any existing guidelines on this. Thanks :)

**Comment by halton** [comment]
@yuankaining the one line of brief message is not a guideline. You could see lots of commits with very detailed information.

**Comment by yuankaining** [comment]
Thanks @halton . I dug a little bit and found the following. There are a number of existing long commit messages :)

    $ git log --log-size | egrep "log size [0-9]{4}" | sort | tail
    log size 2969
    log size 3185
    log size 3351
    log size 3422
    log size 3735
    log size 3836
    log size 4244
    log size 4347
    log size 6326
    log size 8155

**Comment by rakuco** [comment]
> I'm following other colleagues' commit message style -- one line of brief message, one line of JIRA bug ID. Is this style welcomed? Let me know if there is any existing guidelines on this. Thanks :)

Well, I've been bugging your colleagues for years to be less concise, with mixed results ;)

We do have a few commit message guidelines in [CONTRIBUTING.md](https://github.com/crosswalk-project/crosswalk/blob/master/CONTRIBUTING.md), but they are not very extensive. [Section 2 of the Linux kernel's SubmittingPatches file](https://git.kernel.org/cgit/linux/kernel/git/torvalds/linux.git/tree/Documentation/SubmittingPatches?id=refs/tags/v4.5-rc2#n106) covers the subject in a lot more detail, and the majority of what's written there applies to Crosswalk commit/pull request messages too.

The rule of thumb is: if in doubt, be verbose. There is no reason why a shorter explanation would be better than a longer explanation. Always assume the people looking at your changes are not deeply familiar with what you are doing, so you need to convince them that your patch is worth integrating by describing what was wrong/broken and why your change is fixing things correctly. Not only that, but in a few years you and I might have moved on to a different project, and a new person might have just started working on Crosswalk. They must be able to look at the commit logs and understand why a certain change was made, without having to browse through mailing lists or go ask someone who was around before about it.

**Pull request merged**